staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,150 @@
+	emulationVersion        VersionVar
+	minCompatibilityVersion VersionVar
+}
+
+func (m *effectiveVersions) BinaryVersion() *version.Version {
+	return m.binaryVersion.Load()
+}
+
+func (m *effectiveVersions) EmulationVersion() *version.Version {
+	return m.emulationVersion.Val.Load()
+}
+
+func (m *effectiveVersions) MinCompatibilityVersion() *version.Version {
+	return m.minCompatibilityVersion.Val.Load()
+}
+
+func (m *effectiveVersions) Set(binaryVersion, emulationVersion, minCompatibilityVersion *version.Version) {
+	m.binaryVersion.Store(binaryVersion)
+	m.emulationVersion.Val.Store(emulationVersion)
+	m.minCompatibilityVersion.Val.Store(minCompatibilityVersion)
+}
+
+func (m *effectiveVersions) SetBinaryVersionForTests(binaryVersion *version.Version) func() {
+	oldBinaryVersion := m.binaryVersion.Load()
+	m.Set(binaryVersion, binaryVersion, version.MajorMinor(binaryVersion.Major(), binaryVersion.SubtractMinor(1).Minor()))
+	return func() {
+		m.Set(oldBinaryVersion, oldBinaryVersion, version.MajorMinor(oldBinaryVersion.Major(), oldBinaryVersion.SubtractMinor(1).Minor()))
+	}
+}
+
+func (m *effectiveVersions) Validate() []error {
+	// Validate only checks the major and minor versions.
+	binaryVersion := m.binaryVersion.Load().WithPatch(0)
+	emulationVersion := m.emulationVersion.Val.Load().WithPatch(0)
+	minCompatibilityVersion := m.minCompatibilityVersion.Val.Load().WithPatch(0)
+
+	maxEmuVer := binaryVersion
+	minEmuVer := binaryVersion
+	// emulationVersion can only be 1.{binaryMinor-1}...1.{binaryMinor} for alpha if EmulationVersion feature gate is enabled.
+	// otherwise, emulationVersion can only be 1.{binaryMinor}.
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.EmulationVersion) {
+		minEmuVer = binaryVersion.SubtractMinor(1)
+	}
+	if emulationVersion.GreaterThan(maxEmuVer) || emulationVersion.LessThan(minEmuVer) {
+		return []error{fmt.Errorf("emulation version %s is not between [%s, %s]", emulationVersion.String(), minEmuVer.String(), maxEmuVer.String())}
+	}
+	// minCompatibilityVersion can only be 1.{binaryMinor-1} for alpha.
+	maxCompVer := binaryVersion.SubtractMinor(1)
+	minCompVer := binaryVersion.SubtractMinor(1)
+	if minCompatibilityVersion.GreaterThan(maxCompVer) || minCompatibilityVersion.LessThan(minCompVer) {
+		return []error{fmt.Errorf("minCompatibilityVersion version %s is not between [%s, %s]", minCompatibilityVersion.String(), minCompVer.String(), maxCompVer.String())}
+	}
+	return []error{}
+}
+
+func (m *effectiveVersions) AddFlags(fs *pflag.FlagSet) {
+	if m == nil {
+		return
+	}
+
+	fs.Var(&m.emulationVersion, "emulation-version", ""+
+		"The version API server emulates its capabilities (APIs, features, ...) of.\n"+
+		"If set, the server would behave like the set version instead of the underlying binary version.\n"+
+		"Defaults to the binary version. Should be between 1.{binaryMinorVersion-1}..{binaryVersion}.\n"+
+		"Format could be semantic or generic version, like v1.30.0-alpha|v1.30|1.30")
```

Comment:
currently `resourceExpirationEvaluator` considers prerelease in its version https://github.com/kubernetes/kubernetes/blob/11785bb815d58eb553be3a1fa305464c35d860cc/staging/src/k8s.io/apiserver/pkg/server/deleted_kinds.go#L37. So I guess we will have to use the semantic format. But on the other hand, patch version does not matter for emulation version. So that's kind of confusing.

Reply:
I think it's OK to use semantic version.  To avoid confusion for users, we only need to add additional validation to this flag to disallow setting of anything but major and minor versions.

I also want to make sure we don't accidentally allow a patch version to be set by users on emulation version and then have some code in k8s checking patch version.  That would be BAD.